### PR TITLE
Refactor the export services to use the new paradigm

### PR DESCRIPTION
**Pull request by brandonforster**
#368.

Does not currently pass tests. There is a panic in a function describing a panic. Could use an assist.

### CODE REVIEW

File: pkg/models/configurationstruct.go [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // SPDX-License-Identifier: Apache-2.0
 //
```
Comment (drasko) [inline pkg/models/configurationstruct.go]: filename long and difficult to read. Please use underscores.

File: pkg/models/deviceidfilter.go [hunk 2]
```diff
@@ -4,27 +4,25 @@
 //
 // SPDX-License-Identifier: Apache-2.0
```
Comment (drasko) [inline pkg/models/deviceidfilter.go]: Filename long ad difficult to read.

File: pkg/models/exportinterfaces.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
+//
+// Copyright (c) 2017
```
Comment (drasko) [inline pkg/models/exportinterfaces.go]: It's 2018.

File: pkg/models/notifyupdate.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package models
```
Comment (drasko) [inline pkg/models/notifyupdate.go]: Add file copyright header.

File: pkg/models/zaplogger.go [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package models
```
Comment (drasko) [inline pkg/models/zaplogger.go]: File header missing.

File: internal/pkg/db/memory/export-memory-database.go [hunk 6]
```diff
@@ -13,33 +13,34 @@
  * or implied. See the License for the specific language governing permissions and limitations under
```
Comment (jduranf) [inline internal/pkg/db/memory/export-memory-database.go]: Maybe the name should be export.go instead of export-memory-database.go, like it's done with core data.go and metadata.go

File: internal/pkg/db/memory/export-memory-database.go [hunk 7]
```diff
@@ -13,33 +13,34 @@
...
-	"github.com/edgexfoundry/edgex-go/export"
 	"gopkg.in/mgo.v2/bson"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
+	"github.com/edgexfoundry/edgex-go/pkg/models"
 )
 
-type memDB struct {
-	regs []export.Registration
+type ExportMemoryDB struct {
+	regs []models.Registration
```
Comment (jduranf) [inline internal/pkg/db/memory/export-memory-database.go]: This line can be moved into client.go, like data and metadata

File: internal/pkg/db/memory/export-memory-database.go [hunk 8]
```diff
@@ -13,33 +13,34 @@
...
 }
 
-func newMemoryClient() *memDB {
-	return &memDB{
-		regs: make([]export.Registration, 0),
+func NewExportMemoryClient() *ExportMemoryDB {
+	return &ExportMemoryDB{
+		regs: make([]models.Registration, 0),
```
Comment (jduranf) [inline internal/pkg/db/memory/export-memory-database.go]: This can be modified in order to work like data and metadata

File: internal/pkg/db/mongo/export-mongo-client.go [hunk 9]
```diff
@@ -32,13 +34,13 @@ Export client client
...
-type MongoClient struct {
+type ExportMongoClient struct {
 	Session  *mgo.Session  // Mongo database session
 	Database *mgo.Database // Mongo database
 }
 
-// Return a pointer to the MongoClient
-func newMongoClient(config DBConfiguration) (*MongoClient, error) {
+// Return a pointer to the ExportMongoClient
+func NewExportMongoClient(config db.Configuration) (*ExportMongoClient, error) {
```
Comment (jduranf) [inline internal/pkg/db/mongo/export-mongo-client.go]: This can be removed and use NewClient() and Connect() defined in internal/pkg/db/mongo/client.go

File: pkg/models/configurationstruct.go (hunk 1 above)
Comment (feclare) [reply pkg/models/configurationstruct.go]: This file should not be moved from export distro. It does not make sense to have it in models.

File: pkg/models/deviceidfilter.go (hunk 2 above)
Comment (feclare) [reply pkg/models/deviceidfilter.go]: This types are local to distro. It does not make any sense moving it to models.

File: pkg/models/encryption.go [hunk 10]
```diff
@@ -4,7 +4,7 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-package distro
```
Comment (feclare) [inline pkg/models/encryption.go]: This types are local to distro. It does not make any sense moving it to models.

File: pkg/models/exportinterfaces.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
+//
```
Comment (feclare) [inline pkg/models/exportinterfaces.go]: This types are local to distro. It does not make any sense moving it to models.

File: pkg/models/filter.go [hunk 12]
```diff
@@ -4,7 +4,7 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-package export
+package models
```
Comment (feclare) [inline pkg/models/filter.go]: This types are local to distro. It does not make any sense moving it to models.

File: pkg/models/httpsender.go [hunk 13]
```diff
@@ -7,14 +7,13 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-package distro
+package models
```
Comment (feclare) [inline pkg/models/httpsender.go]: This types are local to distro. It does not make any sense moving it to models.

File: pkg/models/influxdb.go [hunk 14]
```diff
@@ -4,14 +4,13 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-package distro
+package models
```
Comment (feclare) [inline pkg/models/influxdb.go]: This types are local to distro. It does not make any sense moving it to models.

File: pkg/models/notifyupdate.go (hunk 4 above)
Comment (feclare) [inline pkg/models/notifyupdate.go]: Why separating this types to a new file. I think it's better to have them in registration.go as they are related.

File: pkg/models/xmpp.go [hunk 15]
```diff
@@ -6,14 +6,13 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-package distro
+package models
```
Comment (feclare) [inline pkg/models/xmpp.go]: This types are local to distro. It does not make any sense moving it to models.

**Comment by brandonforster** [comment]
I did the best I could, but build is currently broken on master. Will attempt to fix broken master later.

**Comment by tsconn23** [comment]
@brandonforster A couple things:

1.) Please rebase the branch from master to clear the "out-of-date" message
2.) When the export-client service is run, it throws an error
```
2018/08/27 19:39:50 http: panic serving 127.0.0.1:59720: runtime error: invalid memory address or nil pointer dereference
goroutine 4 [running]:
net/http.(*conn).serve.func1(0xc4201ba5a0)
	/usr/local/go/src/net/http/server.go:1697 +0xd0
panic(0x1370680, 0x15d0400)
	/usr/local/go/src/runtime/panic.go:491 +0x283
github.com/edgexfoundry/edgex-go/internal/export/client.getAllReg(0x15a6280, 0xc4201cc000, 0xc4201c8000)
	/Users/tconn/go/src/github.com/edgexfoundry/edgex-go/internal/export/client/registration.go:82 +0x53
```
This appears to be because the "dbc" variable on line 82 of registration.go is not initialized.

File: internal/export/client/init.go [hunk 16]
```diff
@@ -55,8 +58,8 @@ func Init(conf ConfigurationStruct, l *zap.Logger) error {
 	var err error
 
 	// Create a database client
-	dbc, err = clients.NewDBClient(clients.DBConfiguration{
-		DbType:       clients.GetDatabaseType(conf.DBType),
+	mongo.NewClient(db.Configuration{
```
Comment (tsconn23) [inline internal/export/client/init.go]: @brandonforster As you can see here, the `dbc, err` on the left side of the assignment was removed. This is most likely the cause of the panic.

File: pkg/models/zaplogger.go [hunk 17]
```diff
@@ -0,0 +1,21 @@
+/*******************************************************************************
```
Comment (feclare) [inline pkg/models/zaplogger.go]: Is this file still needed after removing the models from this directory?

File: internal/export/distro/client.go [hunk 18]
```diff
@@ -26,20 +26,20 @@ func getRegistrationBaseURL(host string) string {
 		"/api/v1/registration"
 }
 
-func getRegistrations() []export.Registration {
+func GetRegistrations() []export.Registration {
```
[the remark this answers is not in the document]
Comment (brandonforster) [reply internal/export/distro/client.go]: At one point I had all the registration functionality together but it quickly became apparent that that would result in an untangleable knot of import cycles. This is an artifact of that and it should not be included. Thanks for the catch.

File: internal/pkg/db/db.go [hunk 19]
```diff
@@ -52,6 +52,7 @@ var (
 )
 
 type Configuration struct {
+	DbType       string
```
[the remark this answers is not in the document]
Comment (brandonforster) [reply internal/pkg/db/db.go]: Export Client uses that field, but looking at its usage we probably don't actually need it.

File: internal/export/distro/httpsender_test.go [hunk 20]
```diff
@@ -15,7 +15,7 @@ import (
 	"net/url"
 	"strconv"
 	"testing"
-
+
```
Comment (drasko) [inline internal/export/distro/httpsender_test.go]: Remove tab.

File: internal/export/distro/registrations.go [hunk 21]
```diff
@@ -15,18 +15,31 @@ package distro
...
+	format       interfaces.Formatter
+	compression  interfaces.Transformer
+	encrypt      interfaces.Transformer
+	sender       interfaces.Sender
+	filter       []interfaces.Filterer
+
+	chRegistration chan *export.Registration
+	chEvent        chan *models.Event
+
+	deleteMe bool
```
Comment (drasko) [inline internal/export/distro/registrations.go]: I would prefer just `delete` or `deleteFlag` for the name of this var. Looks confusing, like the var itself should be deleted :).

**Comment by drasko** [comment]
Good one. @feclare please review.

File: internal/export/client/init.go [hunk 22]
```diff
@@ -55,15 +58,19 @@ func Init(conf ConfigurationStruct, l *zap.Logger) error {
 	var err error
 
 	// Create a database client
-	dbc, err = clients.NewDBClient(clients.DBConfiguration{
-		DbType:       clients.GetDatabaseType(conf.DBType),
+	dbc = mongo.NewClient(db.Configuration{
```
Comment (feclare) [inline internal/export/client/init.go]: You are hardcoding the db to mongo. Memorydb was also supported.

File: internal/export/client/init_test.go [hunk 23]
```diff
@@ -0,0 +1,23 @@
+//
```
Comment (feclare) [inline internal/export/client/init_test.go]: This test does not make any sense now that memory db is moved out of this package.

File: internal/export/distro/client.go [hunk 24]
```diff
@@ -26,42 +26,43 @@ func getRegistrationBaseURL(host string) string {
...
 		} else {
 			logger.Warn("Could not validate registration", zap.Error(err))
 		}
 	}
-	return results
+	return results, nil
 }
 
-func getRegistrationByName(name string) *export.Registration {
+func GetRegistrationByName(name string) *export.Registration {
```
Comment (feclare) [inline internal/export/distro/client.go]: Why does it need to be public?

File: internal/export/distro/client.go [hunk 25]
```diff
@@ -26,42 +26,43 @@ func getRegistrationBaseURL(host string) string {
...
 	url := getRegistrationBaseURL(configuration.ClientHost) + "/name/" + name
-	return getRegistrationByNameURL(url)
+	return GetRegistrationByNameURL(url)
 }
 
-func getRegistrationByNameURL(url string) *export.Registration {
+func GetRegistrationByNameURL(url string) *export.Registration {
```
Comment (feclare) [inline internal/export/distro/client.go]: Why does it need to be public?

File: internal/export/distro/registrations.go [hunk 26]
```diff
@@ -249,17 +262,11 @@ func Loop(errChan chan error, eventCh chan *models.Event) {
...
-	for allRegs == nil {
-		logger.Info("Waiting for client microservice")
-		select {
-		case e := <-errChan:
-			logger.Info("exit msg", zap.Error(e))
-			return
-		case <-time.After(time.Second):
-		}
-		allRegs = getRegistrations()
+	if err != nil {
```
Comment (feclare) [inline internal/export/distro/registrations.go]: export distro needs export client to be running to start. It is not perfect, but until we decide how are we going to solve this problem across all the microservices this should not be removed.

File: internal/export/interfaces/types.go [hunk 27]
```diff
@@ -1,18 +1,15 @@
...
+// Copyright (c) 2018
 // Cavium
 // Mainflux
 // IOTech
 //
 // SPDX-License-Identifier: Apache-2.0
 //
 
-package distro
+package interfaces
```
Comment (feclare) [inline internal/export/interfaces/types.go]: Why moving the interfaces to a new package? This interfaces are internal to export distro. No one else package are using them.
I should keep them in distro. And also should make the interfaces lowercase.

File: internal/export/registration_test.go [hunk 28]
```diff
@@ -25,7 +27,7 @@ func TestRegistrationValid(t *testing.T) {
 		{"withoutName", "", CompZip, FormatJSON, DestMQTT, EncAes, false},
 		{"wrongCompresion", "reg", "INVALID", FormatJSON, DestMQTT, EncAes, false},
 		{"wrongFormat", "reg", CompZip, "INVALID", DestMQTT, EncAes, false},
-		{"wrongDestination", "reg", CompZip, FormatJSON, "INVALID", EncAes, false},
+		{"wrongDestination", "reg", CompZip, FormatJSON, "INVALID",  EncAes, false},
```
Comment (feclare) [inline internal/export/registration_test.go]: Run gofmt please.

File: internal/export/zaplogger.go [hunk 29]
```diff
@@ -0,0 +1,21 @@
+/*******************************************************************************
```
Comment (feclare) [inline internal/export/zaplogger.go]: Is needed zap logger in this package?

File: internal/pkg/db/memory/export_test.go [hunk 30]
```diff
@@ -4,24 +4,22 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-package clients
+package memory
```
Comment (feclare) [inline internal/pkg/db/memory/export_test.go]: This test should live in edgex-go/internal/pkg/db/test
You can look how it has been done for data and metadata. This way we can reuse this test for all the db implementations.

File: internal/export/client/init.go (hunk 22 above)
Comment (jduranf) [reply internal/export/client/init.go]: Also Redis will be added

File: internal/export/interfaces/types.go (hunk 27 above)
Comment (feclare) [reply internal/export/interfaces/types.go]: It is because encryption.go in internal/export should be moved to internal/export/distro. Once it is moved the interfaces are not used anymore outside of distro.

File: internal/export/zaplogger.go (hunk 29 above)
Comment (feclare) [reply internal/export/zaplogger.go]: Not once encryption.go is moved to distro

**Comment by drasko** [comment]
Merged! Thanks @brandonforster.
